azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_mqtt.py

```diff
@@ -53,12 +58,57 @@ def _cancel_pending_connection_op(self):
             # NOTE: This code path should NOT execute in normal flow. There should never already be a pending
             # connection op when another is added, due to the SerializeConnectOps stage.
             # If this block does execute, there is a bug in the codebase.
-            error = pipeline_exceptions.OperationCancelled(
-                "Cancelling because new ConnectOperation, DisconnectOperation, or ReauthorizeConnectionOperation was issued"
-            )  # TODO: should this actually somehow cancel the operation?
+            if not error:
+                error = pipeline_exceptions.OperationCancelled(
+                    "Cancelling because new ConnectOperation, DisconnectOperation, or ReauthorizeConnectionOperation was issued"
+                )
+            self._cancel_connection_watchdog(op)
             op.complete(error=error)
             self._pending_connection_op = None
 
+    @pipeline_thread.runs_on_pipeline_thread
+    def _start_connection_watchdog(self, connection_op):
+        logger.debug("{}({}): Starting watchdog".format(self.name, connection_op.name))
+
+        self_weakref = weakref.ref(self)
+        op_weakref = weakref.ref(connection_op)
+
+        @pipeline_thread.invoke_on_pipeline_thread
+        def watchdog_function():
+            this = self_weakref()
+            op = op_weakref()
+            if this and op and this._pending_connection_op is op:
+                logger.info(
+                    "{}({}): Connection watchdog expired.  Cancelling op".format(this.name, op.name)
+                )
+                this.transport.disconnect()
+                if this.pipeline_root.connected:
+                    logger.info(
+                        "{}({}): Pipeline is still connected on watchdog expiration.  Sending DisconnectedEvent".format(
+                            this.name, op.name
+                        )
+                    )
+                    this.send_event_up(pipeline_events_base.DisconnectedEvent())
+                this._cancel_pending_connection_op(
+                    error=pipeline_exceptions.OperationCancelled(
+                        "Transport timeout on connection operation"
+                    )
+                )
+
+        connection_op.watchdog = threading.Timer(WATCHDOG_INTERVAL, watchdog_function)
+        connection_op.watchdog.daemon = True
+        connection_op.watchdog.start()
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _cancel_connection_watchdog(self, op):
+        try:
+            if op.watchdog:
+                logger.debug("{}({}): cancelling watchdog".format(self.name, op.name))
+                op.watchdog.cancel()
```

Comment:
>watchdog [](start = 19, length = 8)

i like the name....but do u want to name it just as watchdog or wathcdog_timer following other conventions ?